<h2> 1. Data Modelling using Linear Regression Model

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import metrics

In [3]:
pd.set_option('display.max_columns', None)

#### Reading in the train (df_dummies.csv) and test (test_df_dummies.csv) files which we cleaned in our data cleaning python notebook.

In [4]:
df = pd.read_csv('df_dummies.csv')
test_df = pd.read_csv('test_df_dummies.csv')

In [5]:
df

id tranc_yearmonth             town         flat_type  \
0        88471      2016-05-01  KALLANG/WHAMPOA            4 ROOM   
1       122598      2012-07-01           BISHAN            5 ROOM   
2       170897      2013-07-01      BUKIT BATOK         EXECUTIVE   
3        86070      2012-04-01           BISHAN            4 ROOM   
4       153632      2017-12-01           YISHUN            4 ROOM   
...        ...             ...              ...               ...   
150629   39814      2020-09-01        WOODLANDS         EXECUTIVE   
150630  147177      2017-06-01      JURONG WEST            5 ROOM   
150631  179087      2020-12-01            BEDOK         EXECUTIVE   
150632   21309      2016-05-01       QUEENSTOWN            3 ROOM   
150633  185947      2018-06-01           YISHUN  MULTI-GENERATION   

              flat_model  lease_commence_date  resale_price  tranc_year  \
0                Model A                 2006      680000.0        2016   
1               Improved                 1987      665000.0        2012   
2              Apartment                 1997      838000.0        2013   
3                Model A                 1992      550000.0        2012   
4             Simplified                 1987      298000.0        2017   
...                  ...                  ...           ...         ...   
150629         Apartment                 1995      625000.0        2020   
150630          Improved                 2001      455000.0        2017   
150631         Apartment                 1993      733000.0        2020   
150632          Improved                 1974      338000.0        2016   
150633  Multi Generation                 1987      765000.0        2018   

        tranc_month  lower  upper  mid                     full_flat_type  \
0                 5     10     12   11                     4 ROOM Model A   
1                 7      7      9    8                    5 ROOM Improved   
2                 7     13     15   14                EXECUTIVE Apartment   
3                 4      1      5    3                     4 ROOM Model A   
4                12      1      3    2                  4 ROOM Simplified   
...             ...    ...    ...  ...                                ...   
150629            9      4      6    5                EXECUTIVE Apartment   
150630            6      4      6    5                    5 ROOM Improved   
150631           12     10     12   11                EXECUTIVE Apartment   
150632            5      7      9    8                    3 ROOM Improved   
150633            6      7      9    8  MULTI-GENERATION Multi Generation   

        floor_area_sqft  price_per_sqft  max_floor_lvl  year_completed  \
0               968.760      701.928238             25            2005   
1              1399.320      475.230826              9            1987   
2              1550.016      540.639580             16            1996   
3              1108.692      496.080066             11            1990   
4               893.412      333.552717              4            1987   
...                 ...             ...            ...             ...   
150629         1733.004      360.645446             12            1985   
150630         1184.040      384.277558             16            1998   
150631         1550.016      472.898344             15            1992   
150632          635.076      532.219766             10            1973   
150633         1765.296      433.355086             12            1987   

       residential commercial market_hawker multistorey_carpark  \
0                Y          N             N                   N   
1                Y          N             N                   N   
2                Y          N             N                   N   
3                Y          Y             N                   N   
4                Y          N             N                   N   
...            ...        ...           ...                 ...   

#### Assigning features we plan to use for our model to 'feature_cols' variable.

In [6]:
feature_cols = ['floor_area_sqft','age_sold','mid','hawker_within_2km','mrt_nearest_distance',
                'town_BEDOK', 'town_BISHAN', 'town_BUKIT BATOK',
                'town_BUKIT MERAH', 'town_BUKIT PANJANG', 'town_BUKIT TIMAH',
                'town_CENTRAL AREA', 'town_CHOA CHU KANG', 'town_CLEMENTI',
                'town_GEYLANG', 'town_HOUGANG', 'town_JURONG EAST', 'town_JURONG WEST',
                'town_KALLANG/WHAMPOA', 'town_MARINE PARADE', 'town_PASIR RIS',
                'town_PUNGGOL', 'town_QUEENSTOWN', 'town_SEMBAWANG', 'town_SENGKANG',
                'town_SERANGOON', 'town_TAMPINES', 'town_TOA PAYOH', 'town_WOODLANDS',
                'town_YISHUN']

#### Store the model training process as a function for easier usage when we explore a different combination of features. We will refer to the RMSE and R2 scores to optimize our model.

In [ ]:
def train_test_rmse(dataframe, feature_cols):
    # Assign features and resale price from training data to X and y, respectively
    X = dataframe[feature_cols]
    y = dataframe.resale_price

    X_train, X_test, y_train, y_test = train_test_split(X, y,random_state=99)

    # Train our model using training data
    linreg = LinearRegression()
    linreg.fit(X_train, y_train)

    # Calculate predictions using our model
    y_pred = linreg.predict(X_test)
    return X_train, X_test, y_train, y_test, y_pred

X_train, X_test, y_train, y_test, y_pred = train_test_rmse(df,feature_cols)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 54852, R2 score: 0.853


In [ ]:
# Calculating null RMSE as the baseline model
y_null = np.zeros_like(y_test, dtype=float)
y_null.fill(y_train.mean())

print(f'Null RMSE: {metrics.root_mean_squared_error(y_test,y_null):.0f}')

Null RMSE: 142924


#### Our initial model has a much lower RMSE score, 54852, than the null RMSE, 142924 which is a positive sign but can be improved.
#### An RMSE score of 54852 means that the performance of the model is able to predict prices up to an error of +/- SGD 54.852K, which is not ideal enough for agents to be used as reference due to the large variance.

In [ ]:
# Testing out by removing towns with lower correlation to resale_price

feature_cols_1 = ['floor_area_sqft','age_sold','mid','hawker_within_2km','mrt_nearest_distance',
                'town_BISHAN', 'town_BUKIT BATOK',
                'town_BUKIT MERAH', 'town_BUKIT TIMAH','town_QUEENSTOWN',
                'town_CENTRAL AREA', 'town_JURONG WEST','town_PASIR RIS', 'town_WOODLANDS','town_YISHUN']

X_train, X_test, y_train, y_test, y_pred = train_test_rmse(df,feature_cols_1)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 64225, R2 score: 0.798


In [ ]:
# Added back removed towns and changed hawker_within_2km to hawker_nearest_distance
feature_cols_2 = ['floor_area_sqft','age_sold','mid','hawker_nearest_distance','mrt_nearest_distance',
                'town_BEDOK', 'town_BISHAN', 'town_BUKIT BATOK',
                'town_BUKIT MERAH', 'town_BUKIT PANJANG', 'town_BUKIT TIMAH',
                'town_CENTRAL AREA', 'town_CHOA CHU KANG', 'town_CLEMENTI',
                'town_GEYLANG', 'town_HOUGANG', 'town_JURONG EAST', 'town_JURONG WEST',
                'town_KALLANG/WHAMPOA', 'town_MARINE PARADE', 'town_PASIR RIS',
                'town_PUNGGOL', 'town_QUEENSTOWN', 'town_SEMBAWANG', 'town_SENGKANG',
                'town_SERANGOON', 'town_TAMPINES', 'town_TOA PAYOH', 'town_WOODLANDS',
                'town_YISHUN']

X_train, X_test, y_train, y_test, y_pred = train_test_rmse(df,feature_cols_2)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 54808, R2 score: 0.853


In [ ]:
# Added in new numerical features- vacancy and cutoff_point from feature_cols_2
feature_cols_3 = ['floor_area_sqft','age_sold','mid','hawker_nearest_distance','mrt_nearest_distance', 'vacancy', 'cutoff_point',
                'town_BEDOK', 'town_BISHAN', 'town_BUKIT BATOK',
                'town_BUKIT MERAH', 'town_BUKIT PANJANG', 'town_BUKIT TIMAH',
                'town_CENTRAL AREA', 'town_CHOA CHU KANG', 'town_CLEMENTI',
                'town_GEYLANG', 'town_HOUGANG', 'town_JURONG EAST', 'town_JURONG WEST',
                'town_KALLANG/WHAMPOA', 'town_MARINE PARADE', 'town_PASIR RIS',
                'town_PUNGGOL', 'town_QUEENSTOWN', 'town_SEMBAWANG', 'town_SENGKANG',
                'town_SERANGOON', 'town_TAMPINES', 'town_TOA PAYOH', 'town_WOODLANDS',
                'town_YISHUN']

X_train, X_test, y_train, y_test, y_pred = train_test_rmse(df,feature_cols_3)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 54748, R2 score: 0.853


In [ ]:
# Added in tranc_year from feature_cols_2 and removed vacancy and cutoff point
feature_cols_4 = ['floor_area_sqft','age_sold','mid','hawker_nearest_distance','mrt_nearest_distance', 'tranc_year',
                'town_BEDOK', 'town_BISHAN', 'town_BUKIT BATOK',
                'town_BUKIT MERAH', 'town_BUKIT PANJANG', 'town_BUKIT TIMAH',
                'town_CENTRAL AREA', 'town_CHOA CHU KANG', 'town_CLEMENTI',
                'town_GEYLANG', 'town_HOUGANG', 'town_JURONG EAST', 'town_JURONG WEST',
                'town_KALLANG/WHAMPOA', 'town_MARINE PARADE', 'town_PASIR RIS',
                'town_PUNGGOL', 'town_QUEENSTOWN', 'town_SEMBAWANG', 'town_SENGKANG',
                'town_SERANGOON', 'town_TAMPINES', 'town_TOA PAYOH', 'town_WOODLANDS',
                'town_YISHUN']

X_train, X_test, y_train, y_test, y_pred = train_test_rmse(df,feature_cols_4)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 54762, R2 score: 0.853


#### After trying a few different combinations, we decided to use feature_cols_4 as our final combination of features and explore other regression models.

<h2> 2. Data Modelling using Decision Tree Regressor Model

In [50]:
from sklearn.tree import DecisionTreeRegressor

In [ ]:
# Assign features and resale price from training data to X and y, respectively
X = df[feature_cols_4]
y = df.resale_price

X_train, X_test, y_train, y_test = train_test_split(X, y,random_state=99)

# Train our model using training data
regressor = DecisionTreeRegressor(random_state=99)
regressor.fit(X_train, y_train)

# Calculate predictions using our model
y_pred = regressor.predict(X_test)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 35332, R2 score: 0.939


In [ ]:
# Re-train the model with full X and y dataset
regressor.fit(X, y)

# Assign features and resale price from test data to X and y, respectively
X = test_df[feature_cols_4]
y = test_df.resale_price

# Calculate predictions using our model
y_pred = regressor.predict(X)

print (f'RMSE score: {metrics.root_mean_squared_error(y, y_pred):.0f}, R2 score: {metrics.r2_score(y, y_pred):.3f}')

RMSE score: 33916, R2 score: 0.944


#### When using the same features and training data as the Linear Regression Model, we are able to obtained a much better performance (lower RMSE, higher R2 score) when using the Decision Tree Model.

In [ ]:
from sklearn.model_selection import GridSearchCV

In [ ]:
# Define the parameter grid to search
param_grid = {'max_depth': [3, 5, 7, None],  
              'min_samples_split': list(range(2,20)),
              'min_samples_leaf': [1, 2, 4, 8]}

# Create a GridSearchCV object
grid_search = GridSearchCV(estimator=DecisionTreeRegressor(random_state=99),
                           param_grid=param_grid,
                           cv=2, 
                           scoring='neg_mean_squared_error', 
                           n_jobs=-1, 
                           verbose=1)

# Use Grid Search to find the best parameters for our model
grid_search.fit(X_train, y_train)

# Obtain optimized parameters
print("\nBest Hyperparameters found by GridSearchCV:")
print(grid_search.best_params_)

# Evaluate the best model
y_pred_tuned = grid_search.best_estimator_.predict(X_test)

print(f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred_tuned):.0f}')
print(f'R2 score: {metrics.r2_score(y_test, y_pred_tuned):.3f}')

Fitting 2 folds for each of 288 candidates, totalling 576 fits


c:\Users\brage\anaconda3\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,



Best Hyperparameters found by GridSearchCV:
{'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 16}
RMSE score: 32691
R2 score: 0.948


#### Using a GridSearchCV function, we are able to fine-tune the model's hyperparameters to obtain even better performance.

In [ ]:
# Re-train the model with full X and y dataset
X = df[feature_cols_4]
y = df.resale_price
grid_search.fit(X, y)

# Assign features and resale price from test data to X and y, respectively
X = test_df[feature_cols_4]
y = test_df.resale_price

# Calculate predictions using our optimized model
y_pred_tuned = grid_search.best_estimator_.predict(X)

print(f'RMSE score: {metrics.root_mean_squared_error(y, y_pred_tuned):.0f}')
print(f'R2 score: {metrics.r2_score(y, y_pred_tuned):.3f}')

Fitting 2 folds for each of 288 candidates, totalling 576 fits


c:\Users\brage\anaconda3\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


RMSE score: 31189
R2 score: 0.953


<h2> 3. Data Modelling using Random Forest Regressor Model

In [55]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
# Assign features and resale price from training data to X and y, respectively
X = df[feature_cols_4]
y = df.resale_price

X_train, X_test, y_train, y_test = train_test_split(X, y,random_state=99)

# Train our model using training data
rf_regressor = RandomForestRegressor(random_state=99)
rf_regressor.fit(X_train, y_train)

# Calculate predictions using our model
y_pred = rf_regressor.predict(X_test)

print (f'RMSE score: {metrics.root_mean_squared_error(y_test, y_pred):.0f}, R2 score: {metrics.r2_score(y_test, y_pred):.3f}')

RMSE score: 26157, R2 score: 0.967


In [ ]:
# Re-train the model with full X and y dataset
rf_regressor.fit(X, y)

# Assign features and resale price from test data to X and y, respectively
X = test_df[feature_cols_4]
y = test_df.resale_price

# Calculate predictions using our model
y_pred = rf_regressor.predict(X)

print (f'RMSE score: {metrics.root_mean_squared_error(y, y_pred):.0f}, R2 score: {metrics.r2_score(y, y_pred):.3f}')

RMSE score: 25818, R2 score: 0.968


#### After experimenting with 3 models, we decided on <b>Random Forest Regressor as our chosen model</b>. This is because it produced the lowest RMSE and highest R2 score when used to predict the resale prices of the testing data.

In [ ]:
# Assign features and resale price from training data to X and y, respectively
X = df[feature_cols_4]
y = df.resale_price

X_train, X_test, y_train, y_test = train_test_split(X, y,random_state=99)

# Define the parameter grid to search
param_grid = {
    'n_estimators': [100],
    'max_depth': [3, 5, 7, None], 
    'min_samples_split': list(range(2,20)),
    'min_samples_leaf': [1, 2, 4, 8]
}

# Create a GridSearchCV object
grid_search = GridSearchCV(estimator=RandomForestRegressor(random_state=99),
                           param_grid=param_grid,
                           cv=2, 
                           scoring='neg_mean_squared_error', 
                           n_jobs=-1, 
                           verbose=1)

# Use Grid Search to find the best parameters for our model
grid_search.fit(X_train, y_train)

# Obtain optimized parameters
print("\nBest Hyperparameters found by GridSearchCV:")
print(grid_search.best_params_)

# Evaluate the best model
y_pred_tuned = grid_search.best_estimator_.predict(X_test)

print(metrics.root_mean_squared_error(y_test, y_pred_tuned))
print(metrics.r2_score(y_test, y_pred_tuned))

Fitting 2 folds for each of 288 candidates, totalling 576 fits


c:\Users\brage\anaconda3\Lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,



Best Hyperparameters found by GridSearchCV:
{'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 3, 'n_estimators': 100}
26103.469249794147
0.9666431176595197


In [ ]:
# Re-train the model with full X and y dataset
X = df[feature_cols_4]
y = df.resale_price
grid_search.fit(X, y)

# # Evaluate the best model
X = test_df[feature_cols_4]
y = test_df.resale_price

y_pred_tuned = grid_search.best_estimator_.predict(X)

print(f'RMSE score: {metrics.root_mean_squared_error(y, y_pred_tuned):.0f}')
print(f'R2 score: {metrics.r2_score(y, y_pred_tuned):.3f}')

RMSE score: 25699
R2 score: 0.968


#### Only minor improvements to performance when we use GridSearchCV to tune our model's hyperparameters.

In [106]:
kaggle_submission = test_df['id']
kaggle_submission = pd.DataFrame(kaggle_submission)
kaggle_submission['Predicted'] = y_pred_tuned
kaggle_submission.to_csv('kaggle_submission.csv', index = False)

#### Export the dataframe containing id and predicted resale prices using our model and exported it out as a CSV file.
#### This is to be used for our kaggle submission to evaluate the score.